flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java

```diff
@@ -306,22 +327,73 @@ void deliverOperatorEventToCoordinator(
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        failureCollection.add(ExceptionHistoryEntry.createGlobal(cause));
+        onFailure(cause);
+    }
+
     /**
      * Updates the execution graph with the given task execution state transition.
      *
      * @param taskExecutionStateTransition taskExecutionStateTransition to update the ExecutionGraph
      *     with
      * @return {@code true} if the update was successful; otherwise {@code false}
      */
-    abstract boolean updateTaskExecutionState(
-            TaskExecutionStateTransition taskExecutionStateTransition);
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
+        final Optional<ExecutionVertexID> maybeExecutionVertexId =
+                executionGraph.findExecutionVertexId(taskExecutionStateTransition.getID());
+        final ExecutionState desiredState = taskExecutionStateTransition.getExecutionState();
+        boolean successfulUpdate = getExecutionGraph().updateState(taskExecutionStateTransition);
+        // We only handle failures for the actual transition into the FAILED state.
+        if (successfulUpdate && desiredState == ExecutionState.FAILED) {
+            // We're sure that the executionVertexId has been found, because we've been able to
+            // update the execution graph.
+            final ExecutionVertexID executionVertexId =
+                    maybeExecutionVertexId.orElseThrow(NoSuchElementException::new);
+            final Execution execution =
+                    executionGraph
+                            .findExecution(executionVertexId)
+                            .orElseThrow(NoSuchElementException::new);
+            if (execution.getFailureInfo().isPresent()) {
+                // Only collect failures that are recorded in the Execution
+                failureCollection.add(
+                        ExceptionHistoryEntry.create(execution, execution.getVertexWithAttempt()));
+            }
+            onFailure(extractErrorOrUseDefault(taskExecutionStateTransition));
+        }
+        return successfulUpdate;
+    }
 
-    /**
-     * Callback which is called once the execution graph reaches a globally terminal state.
-     *
-     * @param globallyTerminalState globally terminal state which the execution graph reached
-     */
-    abstract void onGloballyTerminalState(JobStatus globallyTerminalState);
+    private Throwable extractErrorOrUseDefault(
+            TaskExecutionStateTransition taskExecutionStateTransition) {
+        return ErrorInfo.handleMissingThrowable(
+                taskExecutionStateTransition.getError(userCodeClassLoader));
+    }
+
+    List<ExceptionHistoryEntry> getFailures() {
+        return failureCollection;
+    }
+
+    private Optional<RootExceptionHistoryEntry> convertFailures(
+            List<ExceptionHistoryEntry> failureCollection) {
+        if (failureCollection.isEmpty()) {
```

Comment:
@dmvk The async thread from the execution graph (in response to job failing and task cancelling) and the thread from adaptive scheduler (new task updates or global failures) could both access the failureCollection simultaneously. I think we need to make make this thread-safe, or am I missing something?

Reply:
That sounds weird, all interactions should happen in the main thread. Do you have a test to reproduce this?

Reply:
Apologies, you are right.